### DISCUSSION TITLE
Fraction-based timestamp

### DISCUSSION THREAD

**Discussion by mzhang28** [ppy/osu#12985] [Ideas]
Timestamps in the editor snap to certain fractional intervals within a measure, so it only makes sense for them to be represented this way as well. As far as I can tell, lazer still implements timestamps with number of seconds since the beginning of the audio. Instead, I'm proposing to represent them with a more complex data structure that allows for fractional intervals to be represented exactly. This could make things easier for mapping tools development, etc.

I'm currently using this method to develop a (hopefully) more accurate hitsound copier for mappers, but I thought it would be nice to get some community feedback on it first and see if it could be of any use to the actual game.

The way it works is by representing a timestamp with discriminated union of either an absolute or relative timestamp, where absolute timestamps are still represented by milliseconds, while relative timestamps depend on a timing point, a measure number, and a fraction representing how far into the measure the timestamp is. Then there are conversion methods from one to the other (the conversion from relative to absolute is more straightforward, the other way around is a bit hacky and I'm not sure if the way I did it is the best)

My implementation of it can be found here: https://docs.rs/libosu/0.0.5/libosu/enum.TimeLocation.html

**Comment by peppy** [comment]
Unless you're stretching or shrinking time, at the point we're already using floating points values for times it's a bit pointless to represent as timing values (we already have enough accuracy).

I guess one of the valid arguments is a more concise representation of time values. Of course there are benefits and cons of this depending on the situation.

**Comment by LastExceed** [comment]
while time during gameplay is indeed measured with decimals, offsets of hitobjects and timingpoints in beatmaps are rounded to full milliseconds according to [the wiki entry about the .osu file format](https://osu.ppy.sh/help/wiki/osu!_File_Formats/Osu_(file_format)) (ms per beat does use decimals though)
![ss 2018-07-06 at 03 02 39](https://user-images.githubusercontent.com/14909211/42382882-de1a5796-8135-11e8-9570-d235567ab93f.png)
Sub-millisecond precision should definitely be a thing with timing windows down to +-11.5ms (300r in osu!mania with OD10 +HardRock) so I'd suggest using floating point values for offsets instead of integers here, but fractions would be overkill.

StepMania features a fraction-based timing system in their simfiles: instead of placing hitobjects at offsets indicated in ms you specify a beat resolution like 1/4 or 1/8 and indicate on which beats there's an object. And while it does bring the best possible timing accuracy as well as easy time streching/shrinking, the supposedly human-readable [simfile format](https://github.com/stepmania/stepmania/wiki/sm) is a nightmare to manually edit, e.g. minimal adjustments like moving a single note 1/16 of a beat in the middle of a 1/4 jumpstream can require you to rewrite the entire section, or even the full simfile in the worst case. and don't even get me started with offbeat notes.

In the end, while sub-millisecond precision is something we should strife for, I don't think we'll ever need mircosecond precision so simply using floating point values for offsets should suffice.
